src/EditorFeatures/Core/GoToDefinition/ISymbolicNavigationService.cs

```diff
@@ -0,0 +1,23 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis.Editor
+{
+    internal interface ISymbolicNavigationService
+    {
+        /// <summary>
+        /// Is this a symbol of your language? If it is, tell us and you can take over the navigation however you would like.
+        /// </summary>
+        Task<bool> TryNavigateToSymbol(ISymbol symbol, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Is this a metadata name of your language? If it is, tell us and you can take over the navigation however you would like.
+        /// </summary>
+        Task<bool> TryNavigateToSymbol(string metaDataName, CancellationToken cancellationToken);
```

Comment:
Referencing @CyrusNajmabadi 's comment on the [original issue](https://github.com/dotnet/roslyn/issues/8105#issuecomment-465277656). Does that address your question?

Reply:
If the intent is this is following JTF patterns then great. As long as the interface is in Roslyn though we should have comments to that effect; the reason being Roslyn historically _hasn't_ followed JTF rules, and this would be a notable exception.

Reply:
I'm unsure how to approach this. For now I followed the example of other code that I could find in the Roslyn solution, just to make it work so I could test. I'm too unfamiliar with how the threading/concurrency works to say/do anything useful.

Very keen to take advice and apply it though.

Reply:
So the Roslyn pattern you might see for us doing a .Wait() on a task is potentially problematic because it means that method under the covers _can't_ jump to the UI thread. The JoinableTaskFactory library instead allows for a different wait method so callers can do that. Roslyn does a regular Wait all the time, partly for historical reasons (the library implementing the other threading rules simply wasn't available to us in the early days) and also because in many Roslyn cases we know the thing we're waiting for has no UI thread dependencies because we're careful in certain areas.

In this case, since opening files and moving carets is very thread affinitized, I'd say we either:

1. Just say this interface doesn't have an async method, and calls the method on the UI thread.
2. We say it's OK to be called off the UI thread, but then we have to use the JoinableTaskFactory.Run() method described to actually call it.

Given this is explicitly trying to call into other VS components, some of which may expect we're using JoinableTaskFactory, I'd say 2 is reasonable.

And oh: we should still follow the rule that async methods are suffixed with "Async". @sharwell might be curious to know the analyzer didn't trigger here...

Reply:
This weekend I'll see if I can get anywhere with trying to apply approach 2.